docs/source/manual/jooq.rst

```diff
@@ -0,0 +1,120 @@
+
+A DataSource_ and jOOQ SQLDialect_ must be bound with Guice for the ThreadLocalJooqModule_ to use.
+
+.. note:: The DSLContextProvider_ must be injected via Guice (rather than manually constructed) for the JooqTransactional_ annotation to work.
+
+Making a database call:
+
+.. code-block:: java
+
+  public class CreateMyModelCommand {
+    private final DSLContextProvider dslContextProvider;
+
+    @Inject
+    CreateMyModelCommand(DSLContextProvider dslContextProvider) {
+      this.dslContextProvider = dslContextProvider;
+    }
+
+    /**
+     * Inserts a new {@link MyModel} into the database.
+     */
+    @JooqTransactional
+    public MyModelRecord execute(String name, String description) {
+      return dslContextProvider.get()
+          .insertInto(MY_MODEL)
+          .set(MY_MODEL.NAME, name)
+          .set(MY_MODEL.DESCRIPTION, description)
+          .returning()
+          .fetchOne();
+    }
+  }
+
+Configuring the required bindings:
+
+.. code-block:: java
+
+  public class JooqConfigurationModule extends AbstractModule {
+  @Override
+    protected void configure() {
+      install(new ThreadLocalJooqModule());
+    }
+
+    /**
+     * Load data source and connection pool information from configuration
+     */
+    @Provides
+    @Singleton
+    DataSource provideDataSource(ImmutableHierarchicalConfiguration configuration) {
+      try {
+        ComboPooledDataSource dataSource = new ComboPooledDataSource();
+        dataSource.setDriverClass(configuration.getString("config.jooq.driver_class"));
+
+        ImmutableHierarchicalConfiguration connectionConfig =
+            configuration.immutableConfigurationAt("config.jooq.connection");
+        dataSource.setJdbcUrl(connectionConfig.getString("url"));
+        dataSource.setUser(connectionConfig.getString("username"));
+        dataSource.setPassword(connectionConfig.getString("password"));
+
+        ImmutableHierarchicalConfiguration c3p0Config =
+            configuration.immutableConfigurationAt("config.jooq.c3p0");
+        dataSource.setMinPoolSize(c3p0Config.getInt("min_size"));
+        dataSource.setMaxPoolSize(c3p0Config.getInt("max_size"));
+        dataSource.setMaxStatementsPerConnection(c3p0Config.getInt("maxStatementsPerConnection"));
+        dataSource.setAcquireIncrement(c3p0Config.getInt("acquire_increment"));
+        dataSource.setTestConnectionOnCheckout(c3p0Config.getBoolean("testConnectionOnCheckout"));
+        dataSource.setTestConnectionOnCheckin(c3p0Config.getBoolean("testConnectionOnCheckin"));
+        dataSource.setMaxIdleTimeExcessConnections(c3p0Config.getInt("maxIdleTimeExcessConnections"));
+
+        return dataSource;
+      } catch (PropertyVetoException e) {
```

Comment:
Could you have the `provideDataSource` throw PropertyVetoException?

Reply:
I think that should work.

Technically Guice recommends against throwing exceptions in `@Provides` methods at all, but with eager singletons issues will be caught at startup time and these exceptions will be wrapped in a `ProvisionException`.

We had been manually wrapping these in `IllegalStateException` for being consistent and maybe providing some clarity that this really shouldn't have happened.

Does anyone else have an opinion on the current way vs calling out `PropertyVetoException` on the method?

Reply:
I don't think guice will let you add it to the method signature. If you want to throw a checked exception from a provider I think you need to use the extension and follow the guidance here: https://github.com/google/guice/wiki/ThrowingProviders

Reply:
I will keep it as is then.